sphinx/transforms/__init__.py

```diff
@@ -327,3 +329,33 @@ def visit_pending_xref(self, node):
     def visit_image(self, node):
         # type: (nodes.Node) -> None
         raise nodes.SkipNode
+
+
+class SphinxSmartQuotes(DocutilsSmartQuotes):  # NOQA
+    """
+    Customized SmartQuotes to avoid transform for some extra node types.
+    """
+    def get_tokens(self, txtnodes):
+        # A generator that yields ``(texttype, nodetext)`` tuples for a list
+        # of "Text" nodes (interface to ``smartquotes.educate_tokens()``).
+
+        texttype = {True: 'literal',  # "literal" text is not changed:
+                    False: 'plain'}
+        for txtnode in txtnodes:
+            nodetype = texttype[isinstance(txtnode.parent,
+                                           (nodes.literal,
+                                            nodes.literal_block,
+                                            addnodes.literal_emphasis,
+                                            addnodes.literal_strong,
+                                            addnodes.desc_signature,
+                                            addnodes.productionlist,
+                                            addnodes.desc_optional,
+                                            addnodes.desc_name,
```

Comment:
I am not sure if "rendered as literal" is correct wording: for example the ``literal_emphasis`` node has docstring:

> Node that behaves like `emphasis`, but further text processors are not applied (e.g. smartypants for HTML output).

in ``addnodes.py``. The addition of ``addnodes.desc_name`` to this list only prevents SmartQuotes text processor action. A priori, it does not prevent other text processors, hence is less strict than ``literal_emphasis``, a priori.

Background:

The ``option`` directive is defined in the standard domain, associated with class  ``Cmdoption``. The options are put into ``addnodes.desc_name`` nodes.

And this node is used by the ``c``, ``cpp``, ``javascript``, ``python`` and ``rst`` domains in the handling of signatures. It appears appropriate to inhibit smart quotes (in particular dash transforms) there.

Side note: the 1.5.6 ``SmartyPantsHTMLTranslator`` had a ``visit_option`` and ``depart_option`` which looks strange indeed because no ``option`` node name was ever registered with Docutils (see ``addnodes.py``). And the ``Cmdoption`` class in the standard domain uses other nodes like ``addnodes.desc_name``. However, smartypants was inhibited via the ``no_smarty += 1`` done by ``visit_desc_signature``. Thus, it looked fine.

With Sphinx 1.6.x and the direct use of the Docutils SmartQuotes transform, there is no ``no_smarty`` flag, and the test is done by the ``get_tokens`` method of the SmartQuotes object. Thus I needed to add there ``addnodes.desc_name``. On the other hand, as said above all uses in Sphinx core are for signatures. And it appears appropriate to inhibit smart quotes in any kind of signature.